File: compiler/bench/peephole1.fs
\ expected: 200000000
\ Peephole optimization - redundant operations

: redundant ( n -- n )
  1+ 1-   \ cancels out
  2 * 2 / \ cancels out (for positive)
  ;

: main
  0 100000000 0 do
    i redundant drop 2 +
  loop . cr ;

File: compiler/bench/peephole1.c
// expected: 200000000
// Peephole optimization - redundant operations
#include <stdio.h>

static long redundant(long n) {
    n = n + 1;
    n = n - 1;  // cancels
    n = n * 2;
    n = n / 2;  // cancels
    return n;
}

int main() {
    long sum = 0;
    for (long i = 0; i < 100000000; i++) {
        redundant(i);
        sum += 2;
    }
    printf("%ld\n", sum);
    return 0;
}
